core/src/main/java/org/elasticsearch/script/ScriptMetaData.java

```diff
@@ -0,0 +1,299 @@
+
+import org.elasticsearch.ResourceNotFoundException;
+import org.elasticsearch.cluster.AbstractDiffable;
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.cluster.DiffableUtils;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.ParseFieldMatcher;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentParser.Token;
+import org.elasticsearch.common.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+public final class ScriptMetaData implements MetaData.Custom {
+
+    public final static String TYPE = "stored_scripts";
+    public final static ScriptMetaData PROTO = new ScriptMetaData(Collections.emptyMap());
+
+    private final Map<String, ScriptAsBytes> scripts;
+
+    ScriptMetaData(Map<String, ScriptAsBytes> scripts) {
+        this.scripts = scripts;
+    }
+
+    public BytesReference getScriptAsBytes(String language, String id) {
+        ScriptAsBytes scriptAsBytes = scripts.get(toKey(language, id));
+        if (scriptAsBytes != null) {
+            return scriptAsBytes.script;
+        } else {
+            return null;
+        }
+    }
+
+    public String getScript(String language, String id) {
+        BytesReference scriptAsBytes = getScriptAsBytes(language, id);
+        if (scriptAsBytes == null) {
+            return null;
+        }
+
+        // Scripts can be stored via API in several ways:
+        // 1) wrapped into a 'script' json object or field
+        // 2) wrapped into a 'template' json object or field
+        // 3) just as is
+        // In order to fetch the actual script in consistent manner this parsing logic is needed:
+        try (XContentParser parser = XContentHelper.createParser(scriptAsBytes);
+             XContentBuilder builder = XContentFactory.contentBuilder(XContentType.JSON)) {
+            parser.nextToken();
+            parser.nextToken();
+            switch (parser.currentName()) {
+                case "script":
+                    if (parser.nextToken() == Token.VALUE_STRING) {
+                        return parser.text();
+                    } else {
+                        builder.copyCurrentStructure(parser);
+                    }
+                    break;
+                case "template":
```

Comment:
Good point. It has been removed: https://github.com/elastic/elasticsearch/commit/22dc06fb7d129bea674b6a59a1b518d481df881e